cleanup static_cast of AutogradMeta

albanD opened the pull request:
The goal is to reduce the spread of static casts in the autograd code as per the comment in https://github.com/pytorch/pytorch/pull/49097#discussion_r543695091
I wasn't sure how to use a virtual method here but a simple method in impl clean it up quite nicely.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54103 cleanup static_cast of AutogradMeta**
* #54102 properly make AutogradMeta/DifferentiableViewMeta attributes internal
* #54101 Remove legacy from optional-related function names
* #54100 Add size check for forward grads
* #54099 make internal forwardAD methods on at::Tensor internal

Differential Revision: [D27117840](https://our.internmc.facebook.com/intern/diff/D27117840)

albanD committed:
cleanup static_cast of AutogradMeta

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -178,19 +177,20 @@ inline Tensor as_view(const Tensor & base, const Tensor & tensor, bool is_bw_dif
-      new_fw_info = base_fw_info.chain(base, tensor, view_func);
+    if (base_meta && base_meta->is_view_) {
+      auto diff_view_meta = torch::autograd::impl::get_view_autograd_meta(base);
+      if (diff_view_meta->has_fw_view()) {
```
Is it intended that the behavior here changed? If this if condition is false, then we will no longer go into the else.

albanD replied:
Good catch.
I wanted to reduce the amount of time we access the autograd meta but that's not really possible I guess.
I'll revert back to the old logic

soulitzer commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -225,22 +225,23 @@ inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor>& ten
-      new_fw_info = ViewInfo(base_fw_info.base_, /* view_func */ nullptr);
+    if (base_meta && base_meta->is_view_) {
+      auto diff_view_meta = torch::autograd::impl::get_view_autograd_meta(base);
+      if (diff_view_meta->has_fw_view()) {
```
Same for here?

soulitzer commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -322,6 +320,12 @@ namespace impl {
     return static_cast<AutogradMeta*>(self.unsafeGetTensorImpl()->autograd_meta());
   }
 
+  DifferentiableViewMeta* get_view_autograd_meta(const Variable& self) {
```
nit: Would we want to use `Tensor` here for our variables names here? or do we keep `Variable` for consistency.

albanD replied:
Yes we have a `using Variable = at::Tensor` at the top of the header.
We should remove that and rename all of them I think yes.

albanD committed:
Update on "cleanup static_cast of AutogradMeta"

The goal is to reduce the spread of static casts in the autograd code as per the comment in https://github.com/pytorch/pytorch/pull/49097#discussion_r543695091
I wasn't sure how to use a virtual method here but a simple method in impl clean it up quite nicely.





[ghstack-poisoned]

ezyang commented:
This seems OK, but given that all the call sites are testing for view-ness before calling the cast, it seems like it might just be better to make this function return a nullptr/ullopt when it's not castable, and then force callers to test the return output.

albanD committed:
Update on "cleanup static_cast of AutogradMeta"

The goal is to reduce the spread of static casts in the autograd code as per the comment in https://github.com/pytorch/pytorch/pull/49097#discussion_r543695091
I wasn't sure how to use a virtual method here but a simple method in impl clean it up quite nicely.



Differential Revision: [D27117840](https://our.internmc.facebook.com/intern/diff/D27117840)

[ghstack-poisoned]

albanD commented:
@ezyang I updated following your recommendation. Does that look better? It does make the code simpler but it is a less clear that we use this to check if a Tensor is a view.

ezyang commented:
One way to make it clearer is to jiggle the naming; e.g., `maybe_view_XXX` or `try_view_XXX`
